### ISSUE TITLE
pwntools 4.6.0 always "Waiting for debugger" when using gdb 10.1

### ISSUE THREAD

**Issue by kokifish** [Gallopsled/pwntools#1984]
I am using pwntools 4.6.0 on kali 2021.3 with gdb 10.1-2.
Once I run `gdb.attach(sh)`, it says "Waiting for debugger" in spite of the gdb terminal has already come out.
It can be resolved after manual downgrade gdb.

gcc version:
gcc --version 
gcc (Debian 10.2.1-6) 10.2.1 20210110
Copyright (C) 2020 Free Software Foundation, Inc.

pwntools version: 
[*] Pwntools v4.6.0

screenshot:
<img width="1285" alt="gdb10_bug_pwntools" src="https://user-images.githubusercontent.com/29633733/136657170-2e3e871b-6aa4-4130-864e-bc55a1600c5d.png">

log:
[log.txt](https://github.com/Gallopsled/pwntools/files/7315530/log.txt)

**Comment by Arusekk** [comment]
What terminal are you using? Try using a different one, e.g. tmux, or specifying your own one in `context.terminal`

**Comment by kokifish** [comment]
> What terminal are you using? Try using a different one, e.g. tmux, or specifying your own one in `context.terminal`

I use zsh, and I found this bug can be resolved by downgrade pwntools to 4.5.1. (with gdb 10.1-2)

**Comment by Arusekk** [comment]
zsh is a shell, mostly irrelevant here; terminal is the GUI (or TUI) that renders the IO of a terminal device; examples: gnome-terminal, alacritty, konsole, xterm, rxvt, tmux...

**Comment by kokifish** [comment]
Thank you for your advice. I use the GNOME Terminal, and the situation is still the same. 
I used the default terminal QTerminal(0.16.1) in kali 2021 before.

> zsh is a shell, mostly irrelevant here; terminal is the GUI (or TUI) that renders the IO of a terminal device; examples: gnome-terminal, alacritty, konsole, xterm, rxvt, tmux...

**Comment by zachriggle** [comment]
@hex-16...

## More Information Needed
* Can you run with `DEBUG` e.g. `python3 yourscript.py DEBUG` and attach the full output or upload to gist.github.com?
* Can you provide your `context.terminal` setting, if set?
* Are you using `gdb.attach()` or `gdb.debug()`?
* When everything is hung at "waiting for debugger", can you `cat /proc/PID/status` of the child process spawned by GDB and attach / upload that as well?
  * You can get the pid from GDB via `(gdb) python print(gdb.selected_inferior().pid)`
  * In particular, we're looking for the `TracerPid:` line, but may as well upload the whole thing

## Troubleshooting
* Notably, I don't see "Wrote gdb script to..." or "running in new terminal" (which the GDB module does), just "Created script for new terminal" (which is done by `run_in_new_terminal`)
* Diffing with `git diff -W -b 4.5.1..4.6.0 -- pwnlib/gdb.py pwnlib/util/misc.py` shows that only a few changes were made to `pwnlib.gdb` in between those releases.
  * Instead of `str` concatenation, we now use a `list` to create the `gdb` command-line.  Should be a noop.
  * The `api=` keyword and RPyc stuff was added.  Not relevant here.
  * Tests were all updated and so the functionality should work, but tests only use `GDB_FAKETERMINAL`.  Possibly relevant.

**Comment by Arusekk** [comment]
This may be in the saga of #1922. Please test whether it works with the newest dev branch.

**Comment by esp0xdeadbeef** [comment]
Got the same setup, versions of software:
Requirement already satisfied: pwntools in /usr/local/lib/python3.9/dist-packages (4.6.0)
Python 3.9.7
$BASH_VERSION = 5.1.8(1)-release
gdb (Debian 10.1-2) 10.1.90.20210103-git



 `python3 exploit.py DEBUG` output:

```bash
[+] Starting local process '/usr/bin/bash' argv=[b'bash'] : pid 33451
[*] running in new terminal: ['/usr/bin/gdb', '-q', '/usr/bin/bash', '33451']
[DEBUG] Created script for new terminal:
    #!/usr/bin/python3
    import os
    os.execve('/usr/bin/gdb', ['/usr/bin/gdb', '-q', '/usr/bin/bash', '33451'], os.environ)
[DEBUG] Launching a new terminal: ['/usr/bin/tmux', 'splitw', '-h', '-F#{pane_pid}', '-P', '/tmp/tmpklnlec5n'][▁] Waiting for debugger
```


cat /proc/PID/status
```
Name:	bash
Umask:	0022
State:	t (tracing stop)
Tgid:	33451
Ngid:	0
Pid:	33451
PPid:	33448
TracerPid:	33453
Uid:	0	0	0	0
Gid:	0	0	0	0
FDSize:	64
Groups:	0 4 20 120 143 
NStgid:	33451
NSpid:	33451
NSpgid:	33451
NSsid:	33451
VmPeak:	    6760 kB
VmSize:	    6760 kB
VmLck:	       0 kB
VmPin:	       0 kB
VmHWM:	    1160 kB
VmRSS:	    1160 kB
RssAnon:	     148 kB
RssFile:	    1012 kB
RssShmem:	       0 kB
VmData:	     276 kB
VmStk:	     132 kB
VmExe:	     748 kB
VmLib:	    1516 kB
VmPTE:	      52 kB
VmSwap:	       0 kB
HugetlbPages:	       0 kB
CoreDumping:	0
THP_enabled:	1
Threads:	1
SigQ:	0/31488
SigPnd:	0000000000000000
ShdPnd:	0000000000000000
SigBlk:	0000000000000000
SigIgn:	0000000000000004
SigCgt:	0000000000010000
CapInh:	0000000000000000
CapPrm:	000001ffffffffff
CapEff:	000001ffffffffff
CapBnd:	000001ffffffffff
CapAmb:	0000000000000000
NoNewPrivs:	0
Seccomp:	0
Seccomp_filters:	0
Speculation_Store_Bypass:	vulnerable
SpeculationIndirectBranch:	always enabled
Cpus_allowed:	ff
Cpus_allowed_list:	0-7
Mems_allowed:	00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000001
Mems_allowed_list:	0
voluntary_ctxt_switches:	11
nonvoluntary_ctxt_switches:	0
```
gdb output 
![image](https://user-images.githubusercontent.com/42157994/138583231-05aa7444-bd29-47fa-8aaf-a666cd7a379f.png)

exploit.py:

```bash
#!/usr/bin/env python3
from pwn import *
context.terminal = ["tmux", "splitw", "-h"]

with process('bash') as p:

    # Attach the debugger
    gdb.attach(p)
    # never comes here...
    p.sendline('whoami')
```

**Comment by Arusekk** [comment]
Fixed by #1902

**Comment by A1-exe** [comment]
> Fixed by #1902

This is still an issue for me. I am using pwntools 4.7.0 and gdb 10.1-2

**Comment by A1-exe** [comment]
> Fixed by #1902

I am still having this issue. I have versions: pwntools 4.7.0, gdb 10.1-2, gcc 11.2.0
Are there any changes I can make locally to fix this?

Inside of tmux I get an infinite yield. Outside of tmux I get this:
![image](https://user-images.githubusercontent.com/28128898/142713507-aa1068c3-f25e-4d4f-8b3d-2210775e7d34.png)



Note: I still get the infinite hang when running the code proposed by OP as well.

**Comment by Arusekk** [comment]
This opens a new tmux window only if you are insude tmux. So no surprise in the second case. But in the first one, what is the debug output?

**Comment by peace-maker** [comment]
Try the beta which includes https://github.com/Gallopsled/pwntools/pull/1922
`pip install -U pwntools==4.8.0b0`

**Comment by A1-exe** [comment]
> Try the beta which includes #1922 `pip install -U pwntools==4.8.0b0`

This fixed the problem. Thanks!

**Comment by re4388** [comment]
Also fix my problem when using the pwntools==4.8.0b0
